hail/python/hail/matrixtable.py

```diff
@@ -2807,25 +2807,205 @@ def write(
+            right_align = row_field_right_align + entry_field_right_align
+
+            # Here we must distinguish between matrix table columns
+            # and the columns that are used to display this matrix table.
+            # Hail displays both the row fields and the matrix table's columns
+            # and entry fields as columns in a tabular format.
+            # We refer to the columns used to display the matrix table as "display columns".
+            display_column_names = None
+            if len(matrix_table.col_key.dtype) == 1 and matrix_table.col_key.dtype[0] in (
+                hl.tstr,
+                hl.tint32,
+                hl.tint64,
+            ):
+                columns = matrix_table.col_key[0].take(n_cols)
+                if len(set(columns)) == len(columns):
+                    display_column_names = truncated_row_fields + [
+                        f'{repr(column)}.{entry_field}' for column in columns for entry_field in entry_fields
+                    ]
+            else:
+                display_column_names = truncated_row_fields + [
+                    f'<col {index}>.{entry_field}' for index in range(0, n_cols) for entry_field in entry_fields
+                ]
+
+            def max_value_width(index):
+                return max(itertools.chain([0], (len(row[index]) for row in truncated_values)))
+
+            display_column_widths = [
+                max(
+                    len(display_column_names[index]),
+                    len(truncated_types[index]) if should_show_types else 0,
+                    max_value_width(index),
+                )
+                for index in range(len(display_column_names))
+            ]
+
+            display_block_slices = []
+            start_index = 0
+            end_index = 1
+            block_width = display_column_widths[0] + 4 if display_column_widths else 0
+            while end_index < len(display_column_names):
+                block_width += display_column_widths[end_index] + 3
+                if block_width > width:
+                    display_block_slices.append(slice(start_index, end_index))
+                    start_index = end_index
+                    block_width = display_column_widths[end_index] + 4
+                end_index += 1
+            display_block_slices.append(slice(start_index, end_index))
+
+            def format_hline(widths):
+                if not widths:
+                    return "++\n"
+                return '+-' + '-+-'.join(['-' * width for width in widths]) + '-+\n'
+
+            def pad(value: str, width: int, right_align: bool):
+                extra_count = width - len(value)
+                if right_align:
+                    return ' ' * extra_count + value
+                else:
+                    return value + ' ' * extra_count
+
+            def format_line(values: List[str], widths: List[int], right_align: List[bool]):
```

Comment:
Prefer including return types on all functions, this one looks like `-> str` to me